lxc/network_forward.go

```diff
@@ -277,38 +281,69 @@ func (c *cmdNetworkForwardCreate) Run(cmd *cobra.Command, args []string) error {
 		forwardPut.Config = map[string]string{}
 	}
 
-	// Get config filters from arguments.
-	for i := 2; i < len(args); i++ {
+	// Get listen address and config from arguments.
+	var listenAddress string
+	for i := 1; i < len(args); i++ {
 		entry := strings.SplitN(args[i], "=", 2)
 		if len(entry) < 2 {
-			return fmt.Errorf(i18n.G("Bad key/value pair: %s"), args[i])
+			// If it's not the first argument it must be a key/value pair.
+			if i != 1 {
+				return fmt.Errorf(i18n.G("Bad key/value pair: %s"), args[i])
+			}
+
+			// Otherwise it is the listen address.
+			listenAddress = args[i]
+			continue
 		}
 
 		forwardPut.Config[entry[0]] = entry[1]
 	}
 
+	if listenAddress != "" && c.flagAllocate != "" {
+		return errors.New("Cannot specify listen address when requesting auto allocation")
+	}
+
+	if listenAddress == "" {
+		if c.flagAllocate == "" {
+			return fmt.Errorf("Must provide a listen address or --allocate=ipv{4,6}")
+		}
+
+		if c.flagAllocate != "ipv4" && c.flagAllocate != "ipv6" {
+			return fmt.Errorf("Invalid --allocate flag %q. Must be one of 'ipv4', or 'ipv6'", c.flagAllocate)
+		}
+
+		if c.flagAllocate == "ipv4" {
+			listenAddress = net.IPv4zero.String()
+		}
+
+		if c.flagAllocate == "ipv6" {
+			listenAddress = net.IPv6zero.String()
+		}
+	}
+
 	// Create the network forward.
 	forward := api.NetworkForwardsPost{
-		ListenAddress:     args[1],
+		ListenAddress:     listenAddress,
 		NetworkForwardPut: forwardPut,
 	}
 
 	forward.Normalise()
 
-	client := resource.server
-
 	// If a target was specified, create the forward on the given member.
 	if c.networkForward.flagTarget != "" {
 		client = client.UseTarget(c.networkForward.flagTarget)
 	}
 
-	err = client.CreateNetworkForward(resource.name, forward)
+	err = client.CreateNetworkForward(networkName, forward)
 	if err != nil {
 		return err
 	}
 
+	parts := strings.Split(transporter.location, "/")
```

Comment:
Need to improve the validation here before you extract the listenAddress to handle unexpected locations.